packages/smart-contracts/src/contracts/ERC20EscrowToPayV1.sol

```diff
@@ -0,0 +1,385 @@
+            "ERC20EscrowToPay: Only payer can call this function."
+        );
+        _;
+    }
+
+    /// Errors
+    error WithdrawFailed(string description);
+
+    /**
+     * @notice Emitted when an new escrow is initiated.
+     * @param paymentReference Reference of the payment related.
+     */
+    event OpenEscrow(bytes indexed paymentReference);
+
+    /**
+     * @notice Emitted when an non-disputed escrow is completed successfully.
+     * @param paymentReference Reference of the payment related.
+     */
+    event EscrowCompleted(bytes indexed paymentReference);
+
+    /**
+     * @notice Emitted when a new dispute is initiated.
+     * @param paymentReference Reference of the payment related.
+     */
+    event DisputeOpened(bytes indexed paymentReference);
+
+    /**
+     * @notice Emitted when a dispute has been successfully resolved.
+     * @param paymentReference Reference of the payment related.
+     */
+    event DisputeResolved(bytes indexed paymentReference);
+
+    /**
+     * @notice Emitted when a lockDisputedFunds function has been executed successfully.
+     * @param paymentReference Reference of the payment related.
+     * @param tokentimelock address of the new tokentimelock contract.
+     * @param paymentToken address of the timelocked ERC20 token.
+     */
+    event LockPeriodStarted(bytes indexed paymentReference, TokenTimelock tokentimelock, IERC20 paymentToken);
+
+    /**
+     * @notice Emitted when a tokentimelock period is completed successfully.
+     * @param paymentReference Reference of the payment related.
+     */
+    event LockPeriodEnded(bytes indexed paymentReference);
+    
+    constructor(address payable _paymentProxyAddress) {
+        owner = payable(msg.sender);
+        paymentProxy = IERC20FeeProxy(_paymentProxyAddress);
+    }
+
+    /// @notice recieve function reverts and returns the funds to the sender.
+    receive() external payable {
+        revert("not payable receive");
+    }
+
+    /** 
+     * @notice Stores the invoice details in struct, then transfers the funds to this Escrow contract.
+     * @param _paymentRef Reference of the payment related.
+     * @param _paymentToken Address of the ERC20 token to deposit.
+     * @param _payee Transfer recipient.
+     * @param _amount Amount to transfer.
+     * @param _feeAmount The amount of fee to pay.
+     * @param _feeAddress The fee recipient.
+     * @dev The fees amount is transfered from the payer's account.
+     * @return success if transaction successeded. 
+     */
+    function openEscrow(
+        bytes memory _paymentRef,
+        IERC20 _paymentToken,
+        uint256 _amount,
+        address _payee,
+        uint256 _feeAmount,
+        address _feeAddress
+    ) 
+        external
+        payable
+        returns (bool success)
+    {
+        require(invoiceMapping[_paymentRef].amount == 0, 
+            "ERC20EscrowToPayV1: This invoice already exists, is this the correct paymentReference?"
```

Comment:
Try to minimize error messages to save gas, we don't do UX at this level (sadly)

Reply:
```suggestion
            "ERC20EscrowToPay: Request already exists."
```

I think this can be version agnostic?